### DISCUSSION TITLE
Reproducibility of the transformation

### DISCUSSION THREAD

**Discussion by maeinss** [TorchIO-project/torchio#1089] [General]
Hi, I am trying to save a RandomElasticDeformation transformation such that it can be applied for future use. The problem I have is I am unable to find a way to do so. Here is a minimal reproducible example of a cube where I tried to see the shape of the deformed cube within a few iterations. (For the given example my aim is to get the same deformed structure with each iteration) Thank you. 

      import numpy as np
      import torch
      import torchio as tio
      import SimpleITK as sitk

      #%% trial3D object
      trialCube = np.zeros((60, 60, 60))
      cube_size = 30 
      x_start = (trialCube.shape[0] - cube_size) // 2
      y_start = (trialCube.shape[1] - cube_size) // 2
      z_start = (trialCube.shape[2] - cube_size) // 2
      trialCube[x_start:x_start+cube_size, y_start:y_start+cube_size, z_start:z_start+cube_size] = 1
      
      #%% transformation
      seed = 1234
      torch.manual_seed(seed)
      transformation = tio.RandomElasticDeformation(
      num_control_points =12, locked_borders=0, 
                                          image_interpolation = "nearest", label_interpolation = "nearest", max_displacement = 7
                                          )
      
      
      #%% created subject
      subject = tio.Subject(
          image = tio.ScalarImage(tensor = np.array([trialCube]))
          )
      subject.plot()

      #%% transformed Matrix and Plotting
      for __ in range(3):
          transformed = transformation(subject)
          transformed.plot()

**Comment by romainVala** [comment]
hi
well, a random transform will by definition give a different (random) set of parameter
beside here you transform the original subject 3 times

`subject= transformation(subject)`

if you want to assign the same deformation at each step then you have to use

`subject.history `and setup and `torchio.ElasticDeformation`

**Reply by maeinss** [reply]
Thank you romainVala, could you help me set up a elastic deformation with the same example, and use subject history to reproduce. Also, is there any way I can set up control points in the regions I want. I apologize I am relatively new to this but it is a very nice tool. 
Update: 
I tried changing the transformation to RandomElastic, but now I have an error doing so. First, it will throw error if I put a single value for control_points or max_displacement so put the values in standard tuple form. 

    transformation = tio.ElasticDeformation(control_points =(2,2,2),
                                  image_interpolation="nearest", label_interpolation="nearest", max_displacement = (2,2,2))
using this trasformation on the subject I get this error now: `AttributeError: 'tuple' object has no attribute 'copy'` which comes from the `get_bspline_transform` line 241.

**Comment by romainVala** [comment]
```
suj_elast = transformation(subject)
suj_elast.plot()

used_transformed = suj_elast.history[0]

for __ in range(2):
  suj_elast = used_transformed(suj_elast)
  suj_elast.plot()

```

history contains the choosen transform (after random parameter choice)
you can set up the transform from the parameter (so you'll understand how to play with it)

```
used_transformed = tio.ElasticDeformation(control_points=suj_elast.history[0].control_points, max_displacement=suj_elast.history[0].max_displacement)
```

so control point is not a tuple, but a grid

**Reply by maeinss** [reply]
Thank you for clarifying, one last thing here for the `control_points`. I am unable to figure out what type of grid structure are we expecting for the `control_points`. In this Class description of `ElasticDeformation`, control points doesn't have description apart from it's a ndarray, thus I assumed it's similar to the  what we have from the the `get_params` function described earlier. But It doesn't work either. Here are the steps:

    class ElasticDeformation(SpatialTransform):
    r"""Apply dense elastic deformation.

    Args:
        control_points:
        max_displacement:
        image_interpolation: See :ref:`Interpolation`.
        label_interpolation: See :ref:`Interpolation`.
        **kwargs: See :class:`~torchio.transforms.Transform` for additional
            keyword arguments.
    """
The get_params function was used to get `control_points` grid for `apply_transforms`

    def get_params(
        num_control_points,
        max_displacement,
        num_locked_borders: int,
    ) -> np.ndarray:
        grid_shape = num_control_points
        num_dimensions = 3
        coarse_field = torch.rand(*grid_shape, num_dimensions)  # [0, 1)
        coarse_field -= 0.5  # [-0.5, 0.5)
        coarse_field *= 2  # [-1, 1]
        for dimension in range(3):
            # [-max_displacement, max_displacement)
            coarse_field[..., dimension] *= max_displacement[dimension]

        # Set displacement to 0 at the borders
        for i in range(num_locked_borders):
            coarse_field[i, :] = 0
            coarse_field[-1 - i, :] = 0
            coarse_field[:, i] = 0
            coarse_field[:, -1 - i] = 0

        return coarse_field.numpy()
    def apply_transform(self, subject: Subject) -> Subject:
        subject.check_consistent_spatial_shape()
        control_points = self.get_params(
            self.num_control_points,  # type: ignore[arg-type]
            self.max_displacement,  # type: ignore[arg-type]
            self.num_locked_borders,
        )
So now I define my control points as from there and try to use it on the transformation. I still get error.

    ctrlPnts = get_params((1,1,1), (2,2,2), 0)  
    transformation = tio.ElasticDeformation(control_points =val,
                              image_interpolation="nearest", label_interpolation="nearest", max_displacement = 2)
  
    suj_elast = transformation(subject) 
Thank you for your help

**Comment by romainVala** [comment]
I do not get any error ...
may be you should replace
`tio.ElasticDeformation(control_points =val,`
by
` tio.ElasticDeformation(control_points =ctrlPnts,`

**Reply by maeinss** [reply]
I wonder why I am having this issue. Sorry about the typo but it is `ctrlPnts` which I used not val. However, I get this error `TypeError: in method 'BSplineTransformInitializer', argument 2 of type 'std::vector< uint32_t,std::allocator< uint32_t > > const &'` when I run the script. The modified script is short and I am pasting it here. Also, I have always wondered is there a way to place control points in the index coordinates you want. Lets say, I want my control points in the area where I assume maximum deformation will occur. Thank you again.

    import numpy as np
    import torch
    import torchio as tio
    import SimpleITK as sitk
    
    #%% trial3D object
    trialCube = np.zeros((60, 60, 60))
    cube_size = 30 
    x_start = (trialCube.shape[0] - cube_size) // 2
    y_start = (trialCube.shape[1] - cube_size) // 2
    z_start = (trialCube.shape[2] - cube_size) // 2
    trialCube[x_start:x_start+cube_size, y_start:y_start+cube_size, z_start:z_start+cube_size] = 1
    
    #%% created subject
    subject = tio.Subject(
        image = tio.ScalarImage(tensor = np.array([trialCube]))
        )
    subject.plot()
    
    #%% Elastic deformation
    #%%
    def get_params(
            num_control_points,
            max_displacement,
            num_locked_borders: int,
        ) -> np.ndarray:
            grid_shape = num_control_points
            num_dimensions = 3
            coarse_field = torch.rand(*grid_shape, num_dimensions)  # [0, 1)
            coarse_field -= 0.5  # [-0.5, 0.5)
            coarse_field *= 2  # [-1, 1]
            for dimension in range(3):
                # [-max_displacement, max_displacement)
                coarse_field[..., dimension] *= max_displacement[dimension]
    
            # Set displacement to 0 at the borders
            for i in range(num_locked_borders):
                coarse_field[i, :] = 0
                coarse_field[-1 - i, :] = 0
                coarse_field[:, i] = 0
                coarse_field[:, -1 - i] = 0
    
            return coarse_field.numpy()
    ctrlPnts = get_params((1,1,1), (2,2,2), 0)  
    transformation = tio.ElasticDeformation(control_points =ctrlPnts,
                                  image_interpolation="nearest", label_interpolation="nearest", max_displacement = (2,2,2))
    #%%
    transformation(subject)

**Reply by fepegar** [reply]
@maeinss from the docs:

> The minimum number of control points is 4 as this transform uses cubic B-splines to interpolate displacement.

**Reply by maeinss** [reply]
Yes, it works now. Thank you. However, ability to specify the control points positions in a structure would be quite a help if there is any way for that. I appreciate your help

**Reply by fepegar** [reply]
What do you mean by "in a structure"?

**Reply by maeinss** [reply]
With structure I meant the 3D volume. In this example it's the `trialCube`. Let's say I want to put 7 control points in the certain index locations of the `trialCube`. At this time, I only have an option to select the number of the control points but not the position of those points about which the deformation will take place. Thank you.

**Reply by fepegar** [reply]
I might be wrong, but I don't think that's possible using SimpleITK.

**Reply by maeinss** [reply]
I understand, thank you for your help.
